src/tools/pkcs11-tool.c

```diff
@@ -3153,9 +3153,61 @@ static int write_object(CK_SESSION_HANDLE session)
 		}
 #endif
 #endif
-	}
-	else
-	if (opt_object_class == CKO_DATA) {
+		break;
+	case CKO_SECRET_KEY:
+		clazz = CKO_SECRET_KEY;
+		type = CKK_AES;
+
+		if (opt_key_type != 0) {
+			if (strncmp(opt_key_type, "AES:", strlen("AES:")) == 0 || strncmp(opt_key_type, "aes:", strlen("aes:")) == 0)
+				type = CKK_AES;
+			else if (strncmp(opt_key_type, "DES3:", strlen("DES3:")) == 0 || strncmp(opt_key_type, "des3:", strlen("des3:")) == 0)
+				type = CKK_DES3;
+			else
+				util_fatal("Unknown key type %s", type);
+		}
+
+		FILL_ATTR(seckey_templ[0], CKA_CLASS, &clazz, sizeof(clazz));
+		FILL_ATTR(seckey_templ[1], CKA_KEY_TYPE, &type, sizeof(type));
+		FILL_ATTR(seckey_templ[2], CKA_TOKEN, &_true, sizeof(_true));
+		FILL_ATTR(seckey_templ[3], CKA_VALUE, &contents, contents_len);
+		n_seckey_attr = 4;
+
+		if (opt_is_private != 0) {
+			FILL_ATTR(seckey_templ[n_seckey_attr], CKA_PRIVATE, &_true, sizeof(_true));
+			n_seckey_attr++;
+		}
+		else {
+			FILL_ATTR(seckey_templ[n_seckey_attr], CKA_PRIVATE, &_false, sizeof(_false));
+			n_seckey_attr++;
+		}
+
+		if (opt_is_sensitive != 0) {
+			FILL_ATTR(seckey_templ[n_seckey_attr], CKA_SENSITIVE, &_true, sizeof(_true));
+			n_seckey_attr++;
+		}
+		else {
+			FILL_ATTR(seckey_templ[n_seckey_attr], CKA_SENSITIVE, &_false, sizeof(_false));
+			n_seckey_attr++;
+		}
+
+		if (opt_object_label != NULL) {
+			FILL_ATTR(seckey_templ[n_seckey_attr], CKA_LABEL, opt_object_label, strlen(opt_object_label));
+			n_seckey_attr++;
+		}
+		if (opt_object_id_len != 0)  {
+			FILL_ATTR(seckey_templ[n_seckey_attr], CKA_ID, opt_object_id, opt_object_id_len);
+			n_seckey_attr++;
+		}
+
+		rv = p11->C_CreateObject(session, seckey_templ, n_seckey_attr, &seckey_obj);
```

Comment:
for consistency, please move this part down to where the other objects are created.